### PR TITLE
Javascript: Implement `ExceptionCollectorListener` and make it default behaviour.

### PR DESCRIPTION
**Pull request by surister**
## Summary of the changes / Why this is an improvement
Implements #30 #36 #37 #34 to the Javascript target

## Checklist

 - [x] Link to issue this PR refers to (if applicable): 
 - [x] [CLA](https://crate.io/community/contribute/cla/) is signed

Also adds support for typescript types, solving #52.

### CODE REVIEW

**Comment by surister** [comment]
The PR might be to big to test, it's 'already tested' with the Python target, I'm more interested of the resulting API, you can quickly check it in https://github.com/crate/cratedb-sqlparse/pull/38/files#diff-111a35f00d7bdf2e15b6a32750f253243ae44df87bcf0465a2fbfd7178db2dc1 @alexdametto

File: cratedb_sqlparse_js/README.md [hunk 1]
```diff
@@ -37,27 +37,161 @@ console.log(queries[0].original_query)
...
+    sqlparse('SELECT COUNT(*) FROM doc.tbl f WHERE .id = 1;', true)
+} catch (e) {
+    console.log(e)
+}
+```
+
+> [!NOTE]
+> It will only raise the first exception it finds, even if you pass in several statements.
+
+### Query metadata.
```
Comment (amotl) [inline cratedb_sqlparse_js/README.md:117]:
```suggestion
### Query metadata
```

File: cratedb_sqlparse_js/README.md [hunk 2]
```diff
@@ -37,27 +37,161 @@ console.log(queries[0].original_query)
...
+
+// Metadata {
+//   tables: [ Table { name: 'tbl12', schema: 'doc' } ],
+//   parameterizedProperties: {},
+//   withProperties: {}
+// }
+
+```
+
+#### Query properties.
```
Comment (amotl) [inline cratedb_sqlparse_js/README.md:136]:
```suggestion
#### Query properties
```

File: cratedb_sqlparse_js/README.md [hunk 3]
```diff
@@ -37,27 +37,161 @@ console.log(queries[0].original_query)
...
+console.log(stmt.metadata.tables)
+// [ Table { name: 'tbl12', schema: 'doc' } ]
+
+table = stmt.metadata.tables[0]
+
+console.log(table.schema, table.name, table.fqn)
+// doc tbl12 "doc"."tbl12"
+```
+
+#### Parameterized properties.
```
Comment (amotl) [inline cratedb_sqlparse_js/README.md:171]:
```suggestion
#### Parameterized properties
```

File: cratedb_sqlparse_js/cratedb_sqlparse/AstBuilder.js [hunk 4]
```diff
@@ -0,0 +1,97 @@
...
+     *
+     * @param {SqlBaseParser.TableNameContext} ctx
+     */
+    visitTableName(ctx) {
+        const fqn = ctx.qname()
+        const parts = this.getText(fqn).split(".")
+
+        let schema = null
+        let name = null;
+        if (parts.length === 1){
```
Comment (amotl) [inline cratedb_sqlparse_js/cratedb_sqlparse/AstBuilder.js:57]:
```suggestion
        if (parts.length === 1) {
```

File: cratedb_sqlparse_js/README.md [hunk 5]
```diff
@@ -37,27 +37,161 @@ console.log(queries[0].original_query)
...
 
 console.log(__cratedb_version__)
-// 5.6.4
+// 5.7.2
 ```
 
 ### Features
-Currently, we support the same features as CrateDB java's parser:
+
+Currently, we support some of the same features as CrateDB java's parser:
```
Comment (amotl) [inline cratedb_sqlparse_js/README.md:52]:
```suggestion
Currently, the parser supports a subset of the features of CrateDB's Java/ANTLR parser:
```

File: cratedb_sqlparse_js/README.md [hunk 6]
```diff
@@ -37,27 +37,161 @@ console.log(queries[0].original_query)
...
+
 - First class CrateDB SQL dialect support.
 - Input is case-insensitive.
-- Native errors as exceptions.
+- Native errors as exceptions or as objects.
 - Dollar strings.
+- Tables
+- Properties and Parametrized properties.
```
Comment (amotl) [inline cratedb_sqlparse_js/README.md:59]:
```suggestion
- Properties and parametrized properties.
```

File: cratedb_sqlparse_js/README.md [hunk 7]
```diff
@@ -37,27 +37,161 @@ console.log(queries[0].original_query)
...
+
+### Exceptions and errors.
+
+By default exceptions are stored in `statement.exception`
```
Comment (amotl) [inline cratedb_sqlparse_js/README.md:63]:
```suggestion
By default, exceptions are stored in `statement.exception`.
```

**Review by amotl — APPROVED** [review]
Thanks a stack! I've just added a few nit suggestions about punctuation, wording, and formatting, to accept at your disposal.

File: cratedb_sqlparse_js/README.md [hunk 8]
```diff
@@ -37,27 +37,161 @@ console.log(queries[0].original_query)
...
+
 - First class CrateDB SQL dialect support.
 - Input is case-insensitive.
-- Native errors as exceptions.
+- Native errors as exceptions or as objects.
 - Dollar strings.
+- Tables
+- Properties and Parametrized properties.
+
+### Exceptions and errors.
```
Comment (amotl) [inline cratedb_sqlparse_js/README.md:61]:
```suggestion
### Exceptions and errors
```

**Comment by surister** [comment]
With the latest commit types generation is added, it works fine on my end.

`installing the package locally in cratedb-alt-admin`:

image

There are still some types, that are not being picked up, I do not know why, for example `Metadata.withProperties`. I set the jsdoc to `object` but `any` is generated. We can improve these on later iterations.

**Review by alexdametto — APPROVED** [review]
> I'm interested on your opinion @alexdametto of the resulting API, you can quickly check it in https://github.com/crate/cratedb-sqlparse/pull/38/files#diff-111a35f00d7bdf2e15b6a32750f253243ae44df87bcf0465a2fbfd7178db2dc1

Amazing job. 💯 I really like it, especially the `raise_exception` and the exception details, these things are definitely needed in https://github.com/crate/cloud/issues/1888. 

Could be possible to add also the line number of where the error occurs? I have seen that we have it inside `stmt.exception.errorMessage` but it would be good to have it stored in a separate field.

**Comment by surister** [comment]
> > I'm interested on your opinion @alexdametto of the resulting API, you can quickly check it in https://github.com/crate/cratedb-sqlparse/pull/38/files#diff-111a35f00d7bdf2e15b6a32750f253243ae44df87bcf0465a2fbfd7178db2dc1
> 
> Amazing job. 💯 I really like it, especially the `raise_exception` and the exception details, these things are definitely needed in [crate/cloud#1888](https://github.com/crate/cloud/issues/1888).
> 
> Could be possible to add also the line number of where the error occurs? I have seen that we have it inside `stmt.exception.errorMessage` but it would be good to have it stored in a separate field.

There is `console.log(stmt.exception.line, stmt.exception.column)`

**Comment by matriv** [comment]
Sorry for the late review, looks good! thx a lot for the effort @surister !

**Pull request merged**